prov/sm2/src/sm2_rma.c

```diff
@@ -0,0 +1,509 @@
+}
+
+void sm2_rma_handle_local_error(struct sm2_ep *ep,
+				struct sm2_xfer_entry *xfer_entry,
+				struct sm2_sar_ctx *ctx, uint64_t err)
+{
+	int ret = 0;
+	if (err == 0)
+		return;
+	ctx->status_flags |= FI_SM2_SAR_ERROR_FLAG;
+
+	if (0 == (ctx->status_flags & FI_SM2_SAR_STATUS_COMPLETED)) {
+		ret = sm2_write_err_comp(ep->util_ep.tx_cq,
+					 (void *) xfer_entry->hdr.context,
+					 xfer_entry->hdr.op_flags, 0,
+					 xfer_entry->hdr.cq_data, err);
+		if (!ret)
+			ctx->status_flags |= FI_SM2_SAR_STATUS_COMPLETED;
+		else
+			FI_WARN(&sm2_prov, FI_LOG_EP_CTRL,
+				"Error in sm2_write_err_comp\n");
+	}
+	sm2_freestack_push(ep, xfer_entry);
+	if (ctx->msgs_in_flight == 0)
+		sm2_free_rma_ctx(ctx);
+}
+
+void sm2_fill_rma_ctx(struct sm2_ep *ep, const struct fi_msg_rma *msg,
+		      uint32_t op, uint64_t op_flags, sm2_gid_t peer_gid,
+		      struct sm2_sar_ctx *ctx)
+{
+	ctx->ep = ep;
+	ctx->msg.context = msg->context;
+	ctx->msg.rma_iov_count = msg->rma_iov_count;
+	ctx->msg.iov_count = msg->iov_count;
+	ctx->msg.addr = msg->addr;
+	ctx->msg.data = msg->data;
+
+	assert(msg->iov_count <= SM2_IOV_LIMIT);
+
+	memset(ctx->msg.desc, 0, msg->iov_count * sizeof(*msg->desc));
+	if (msg->desc) {
+		memcpy(ctx->msg.desc, &msg->desc,
+		       msg->iov_count * sizeof(*msg->desc));
+	}
+	memcpy(ctx->msg.msg_iov, msg->msg_iov,
+	       msg->iov_count * sizeof(*msg->msg_iov));
+	memcpy(ctx->msg.rma_iov, msg->rma_iov,
+	       msg->rma_iov_count * sizeof(*msg->rma_iov));
+
+	ctx->bytes_sent = 0;
+	ctx->bytes_acked = 0;
+	ctx->bytes_total = ofi_total_iov_len(msg->msg_iov, msg->iov_count);
+	ctx->msgs_in_flight = 0;
+	ctx->peer_gid = peer_gid;
+	ctx->op = op;
+	ctx->op_flags = op_flags;
+	ctx->status_flags = 0;
+}
+
+ssize_t sm2_rma_cmd_fill_sar_xfer(struct sm2_xfer_entry *xfer_entry,
+				  struct sm2_sar_ctx *ctx)
+{
+	ssize_t bytes_to_send, payload_used;
+	int ret;
+	void *src, *dest;
+	enum fi_hmem_iface iface;
+	uint64_t device;
+	struct sm2_rma_msg *msg;
+	struct sm2_cmd_sar_rma_msg *cmd_rma;
+	int jv;
+
+	msg = &ctx->msg;
+
+	xfer_entry->hdr.proto = sm2_proto_sar;
+	sm2_generic_format(xfer_entry, ctx->ep->gid, ctx->op, 0, msg->data,
+			   ctx->op_flags, msg->context);
+	cmd_rma = (struct sm2_cmd_sar_rma_msg *) xfer_entry->user_data;
+	memset(cmd_rma, 0, sizeof(*cmd_rma));
+	cmd_rma->sar_hdr.context = ctx;
+
+	payload_used = 0;
+	if (ctx->op == ofi_op_read_req) {
+		cmd_rma->sar_hdr.request_offset = ctx->bytes_requested;
+		ctx->bytes_requested += SM2_RMA_INJECT_SIZE;
+		if (ctx->bytes_requested >= ctx->bytes_total) {
+			ctx->bytes_requested = ctx->bytes_total;
+			cmd_rma->sar_hdr.proto_flags |=
+				FI_SM2_SAR_LAST_MESSAGE_FLAG;
+		}
+		memcpy(cmd_rma->rma_iov, msg->rma_iov,
+		       sizeof(cmd_rma->rma_iov));
+		ctx->msgs_in_flight++;
+
+		return FI_SUCCESS;
+	}
+
+	for (jv = 0; jv < SM2_IOV_LIMIT; jv++) {
+		if (ctx->bytes_sent == ctx->bytes_total)
+			break;
+		bytes_to_send =
+			MIN(msg->msg_iov[0].iov_len, msg->rma_iov[0].len);
```

Comment:
I'll rework this.  I think you are right that it's not quite correct.  In cases where there are non-overlapping remote and local iovs we may have empty space left in our packets.

The zero-indexing is intentional though, because we are calling `ofi_consume_*` which will eat some of the iov and leave `msg_iov[0]` as the next data to send.

Reply:
resolved.